src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -1,913 +1,1138 @@
-                return baseType.AdditionalPropertiesType.Type;
-            }
-
-            var availableProperties = baseType.Properties.Values
-                .Where(p => !p.Flags.HasFlag(TypePropertyFlags.WriteOnly))
-                .Select(p => p.Name)
-                .OrderBy(x => x);
-
-            var diagnosticBuilder = DiagnosticBuilder.ForPosition(propertyExpressionPositionable);
-            var shouldWarn = TypeValidator.ShouldWarn(baseType);
-
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Extensions;
+using Bicep.Core.Parser;
+using Bicep.Core.Resources;
+using Bicep.Core.SemanticModel;
+using Bicep.Core.Syntax;
+using Bicep.Core.Syntax.Visitors;
+using Bicep.Core.Text;
+
+namespace Bicep.Core.TypeSystem
+{
+    public sealed class TypeAssignmentVisitor : SyntaxVisitor
+    {
+        private class TypeAssignment
+        {
+            public TypeAssignment(ITypeReference reference)
+                : this(reference, Enumerable.Empty<Diagnostic>())
+            {
+            }
+
+            public TypeAssignment(ITypeReference reference, IEnumerable<Diagnostic> diagnostics)
+            {
+                Reference = reference;
+                Diagnostics = diagnostics;
+            }
+
+            public ITypeReference Reference { get; }
+
+            public IEnumerable<Diagnostic> Diagnostics { get; }
+        }
+
+        private readonly IResourceTypeProvider resourceTypeProvider;
+        private readonly TypeManager typeManager;
+        private readonly IReadOnlyDictionary<SyntaxBase, Symbol> bindings;
+        private readonly IReadOnlyDictionary<SyntaxBase, ImmutableArray<DeclaredSymbol>> cyclesBySyntax;
+        private readonly IDictionary<SyntaxBase, TypeAssignment> assignedTypes;
+        private readonly IDictionary<SyntaxBase, TypeAssignment> declaredTypes;
+        private readonly SyntaxHierarchy hierarchy;
+
+        public TypeAssignmentVisitor(
+            IResourceTypeProvider resourceTypeProvider,
+            TypeManager typeManager,
+            IReadOnlyDictionary<SyntaxBase, Symbol> bindings,
+            IReadOnlyDictionary<SyntaxBase, ImmutableArray<DeclaredSymbol>> cyclesBySyntax,
+            SyntaxHierarchy hierarchy)
+        {
+            this.resourceTypeProvider = resourceTypeProvider;
+            this.typeManager = typeManager;
+            // bindings will be modified by name binding after this object is created
+            // so we can't make an immutable copy here
+            // (using the IReadOnlyDictionary to prevent accidental mutation)
+            this.bindings = bindings;
+            this.cyclesBySyntax = cyclesBySyntax;
+            this.assignedTypes = new Dictionary<SyntaxBase, TypeAssignment>();
+            this.declaredTypes = new Dictionary<SyntaxBase, TypeAssignment>();

```

Comment:
In the discussion we had with Anders, I thought the suggestion was to only calculate the declared type for a specific node on-demand to avoid the upfront calculation. Is that a concern here?

Reply:
He had that comment for both types of types and that's the plan for another PR. (Although we are currently requesting types for all nodes in the language server by asking for all diagnostics when files change. I'm also pretty sure there's a bug with assigned types being inconsistent depending on when you request it.)